src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,697 @@
+                        }
+                        break;
+                    case ReceiverType::Orchard:
+                        if (params.GetConsensus().NetworkUpgradeActive(anchorHeight, Consensus::UPGRADE_NU5)
+                            && (!spendable.orchardNoteMetadata.empty() || strategy.AllowRevealedAmounts())) {
+                            result.insert(OutputPool::Orchard);
+                        }
+                        break;
+                }
+            }
+            return result;
+        };
+
+        changeAddr = std::visit(match {
+            [&](const CKeyID& keyId) -> ChangeAddress {
+                auto sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                        sendFromAccount,
+                        getAllowedChangePools({ReceiverType::P2PKH}));
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                return sendTo.value();
+            },
+            [&](const CScriptID& scriptId) -> ChangeAddress {
+                auto sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                        sendFromAccount,
+                        getAllowedChangePools({ReceiverType::P2SH}));
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                return sendTo.value();
+            },
+            [](const libzcash::SproutPaymentAddress& addr) -> ChangeAddress {
+                // for Sprout, we return change to the originating address using the tx builder.
+                return addr;
+            },
+            [](const libzcash::SproutViewingKey& vk) -> ChangeAddress {
+                // for Sprout, we return change to the originating address using the tx builder.
+                return vk.address();
+            },
+            [&](const libzcash::SaplingPaymentAddress& addr) -> ChangeAddress {
+                // for Sapling, if using a legacy address, return change to the
+                // originating address; otherwise return it to the Sapling internal
+                // address corresponding to the UFVK.
+                std::optional<RecipientAddress> sendTo = sendFromAccount == ZCASH_LEGACY_ACCOUNT
+                    ? addr
+                    : pwalletMain->GenerateChangeAddressForAccount(
+                            sendFromAccount,
+                            getAllowedChangePools({ReceiverType::Sapling}));
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                return sendTo.value();
+            },
+            [&](const libzcash::SaplingExtendedFullViewingKey& fvk) -> ChangeAddress {
```

Comment:
The logic here is the same as for `SaplingPaymentAddress` but with `fvk.DefaultAddress()` in place of `addr`. Similarly, the logic for `UnifiedFullViewingKey` is the same as for `UnifiedAddress` after resolving to a `ZcashdUnifiedFullViewingKey`. Can we reduce code duplication by either:
* defining closures for the common Sapling parts and the common Orchard parts; or
* resolving to a smaller number of cases by adding another match preceding this one?

Reply:
I did the former, at least partially because I don’t know a coherent name for
```c++
std::variant<
    std::monostate, // transparent pool
    SproutPaymentAddress,
    SaplingPaymentAddress,
    ZcashdUnifiedFullViewingKey,
    AccountId>
```